statechange-led: Also manage locate led for zed vdev events

ZNikke opened the pull request:
Currently, disks being offline for various reasons (manually for disk
replacement, operator mistakes, odd failure modes) have no led
indication on enclosures.

Augment zfs/zed led handling by leveraging the locate led to indicate
offline devices.

<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

A rather common (for us) use case is disks being manually offlined for replacement, in these cases no fault led indicates which disk to replace.  Also, disks can end up being offline due to failure or operator mistakes.

Currently there is no led indication on enclosures to help admins out, and overloading the fault led didn't really make sense when the locate led is currently not used.

Thus we suggest to make use of the locate led by lighting it for all offline devices. It provides a visual difference compared to the error conditions that sets of the fault led, while at the same time provide guidance in operations and error states where disks ends up in an offline state instead of an error state.

### Description
<!--- Describe your changes in detail -->

* Add argument for led instead of hard-coding "fault" in state_to_val()
* Iterate over fault and locate leds where suitable
* Update comments to match function

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

`zpool offline/online` use case tested.

`zpool import` use case not tested yet due to various issues in my 2.1.0-rc4 build that's fixed in master.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

nabijaczleweli commented on cmd/zed/zed.d/statechange-led.sh:
```diff
@@ -94,15 +99,27 @@ check_and_set_led()
+		fi
+	else
+		zed_log_msg "$0 state_to_val error: led=$led not understood"
+	fi
```
This might be better served as something like
```sh
case "$led" in
	fault)
		case "$state" in
			FAULTED|DEGRADED|UNAVAIL) echo 1 ;;
			ONLINE)                   echo 0 ;;
		esac
		;;
	locate)
		case "$state" in
			OFFLINE) echo 1 ;;
			*)       echo 0 ;;
		esac
		;;
	*)
		zed_log_msg "$0 state_to_val error: led=$led not understood"
		;;
esac
```

ZNikke replied:
In a271946d724ae1d399564ace759fce8e00686812 - but superseeded by 138c8d24066bc228ccfa31a3d7a1d30629bdf2e8

nabijaczleweli commented on cmd/zed/zed.d/statechange-led.sh:
```diff
@@ -1,9 +1,14 @@
+#
+# Turn locate LED:
+#  ON  if vdev OFFLINE
+#  OFF if vdev ONLINE
```
This is aligned differently (1 space) than the list above (2 spaces)

ZNikke replied:
In a271946 - but superseeded by 138c8d2

nabijaczleweli commented on cmd/zed/zed.d/statechange-led.sh:
```diff
@@ -1,9 +1,14 @@
-# Turn its LED off when it's back ONLINE again.
+# Turn fault LED:
+#   ON  if vdev FAULTED, DEGRADED or UNAVAIL.
+#   OFF if vdev ONLINE
```
might be best to add an explicit "IGNORE all other states" here, too

ZNikke replied:
In a271946 - but superseeded by 138c8d2

nabijaczleweli commented on cmd/zed/zed.d/statechange-led.sh:
```diff
@@ -122,54 +139,66 @@ process_pool()
-	val=$(state_to_val "$ZEVENT_VDEV_STATE_STR")
 	vdev=$(basename "$ZEVENT_VDEV_PATH")
-	check_and_set_led "$ZEVENT_VDEV_ENC_SYSFS_PATH/fault" "$val"
+	for encled in fault locate ; do
```
space before semicolon

ZNikke replied:
In a271946d724ae1d399564ace759fce8e00686812

nabijaczleweli commented on cmd/zed/zed.d/statechange-led.sh:
```diff
@@ -122,54 +139,66 @@ process_pool()
-		if [ "$current_val" = "$val" ] ; then
-			# LED is already set correctly
-			continue
+		# Treat all non-zero values as 1
```
The entire hunk up to here could be better served as 
```sh
	# The output will be the vdevs only (by matching "/dev/"):
	#
	#    U45     ONLINE       0     0     0   /dev/sdk          0       0
	#    U46     ONLINE       0     0     0   /dev/sdm          0       0
	#    U47     ONLINE       0     0     0   /dev/sdn          0       0
	#    U50     ONLINE       0     0     0   Potential info message /dev/sdbn          0       0
	#
	ZPOOL_SCRIPTS_AS_ROOT=1 $ZPOOL status -c upath,fault_led,locate_led "$pool" | awk '/\/dev\// {print $1 " " $2 " " $(NF-2) " " $(NF-1) " " $NF}' | (
	rc=0
	while read -r vdev state dev fault_val locate_val; do
		# Read out current LED values and path
		# Get dev name (like 'sda')
		dev=$(basename "$dev")
		vdev_enc_sysfs_path=$(realpath "/sys/class/block/$dev/device/enclosure_device"*)
		if [ -z "$vdev_enc_sysfs_path" ] ; then
			# Skip anything with no sysfs LED entries
			continue
		fi

		# Treat all non-zero values as 1
		...
```

This avoids an AWK per variable

ZNikke replied:
In a271946d724ae1d399564ace759fce8e00686812 but superseeded by e561d8d0ec365fc02b78b22388041ead081a5957

nabijaczleweli commented:
To test the pool import without actually importing or exporting your pools, you should be able to run `ZED_ZEDLET_DIR=/etc/zfs/zed.d ZPOOL=/sbin/zpool ZEVENT_POOL_GUID="$(zpool get guid -Ho value yourpool)" /etc/zfs/zed.d/statechange-led.sh`

ZNikke commented:
Thanks for the feedback @nabijaczleweli , I'll implement the suggestions and update the PR. I've not mastered this workflow and all the github features yet, but I'll get there eventually...

nabijaczleweli commented on cmd/zed/zed.d/statechange-led.sh:
```diff
@@ -94,13 +98,44 @@ check_and_set_led()
+			locate=1
+			;;
+		*)
+			zed_log_msg "$0 state_to_val state='$state': LED mapping unknown"
```
these were empty on purpose: no echo is an explicit "don't care", and there are many valid states not listed here, which my version preserved, but this breaks

(also, I find this horrendous readability-wise, but)

ZNikke replied:
My main goal with 138c8d24066bc228ccfa31a3d7a1d30629bdf2e8 is to make a clearer mapping between led status and vdev state. This is all a work in progress, which is why I'm leaving all commits in order to be able to track/backtrack until we can reach some agreement of what's best.

The logging is thought to be a fallback if a state is added to core zfs but not to the script, if it's logged sooner or later someone will discover it and assess whether it's relevant to react to or not.

The current list of states is what I found in the zpoolconcepts man page, if there are many more this might need to be done a bit differently.

ZNikke commented:
Side thought, doesn't `statechange-led.sh` need to react on a vdev being removed/exported and clear any leds lit?

From what I understand of `statechange-led.sh` in master it only clears the fault led if the vdev comes ONLINE, but it doesn't pass that state if a vdev is removed in some way, right? So that should mean that whatever leds are on keeps being lit until manually turned off or that disk slot gets used by zfs again...

ZNikke commented:
> Some initial thoughts before I look at the code:
> 
>     1. Have you tested this with REMOVED/UNAVAIL vdevs?  I seem to remember we had an enclosure problem once where some of our vdevs were showing up as UNAVAIL, and as a result they didn't have `vdev_enc_sysfs_path` set.  That behavior makes sense, since if a flaky enclosure doesn't think a disk is in the slot, it's not going to create the disk's `/sys/class/block/sd*/device/enclosure_device*` symlink that we need to set the LEDs.

Not on real hardware by yanking/pulling disks. Due to issue #11950 it's kinda hard when enclosure paths are wrong to begin with... But at least for the REMOVED use case with a sysadmin yanking wrong drive it should work as the enc path should be there.

FWIW, we've had more UNAVAIL's due to broken drives than flaky enclosures, but that's just different (bad) luck I guess...

>     2. You'll want to make the locate LED behavior configurable so operators can turn it off if they just want the fault LED behavior.  Right now we have this in `zed.rc`:
> 
> ```
> ##
> # Turn on/off enclosure LEDs when drives get DEGRADED/FAULTED.  This works for
> # device mapper and multipath devices as well.  Your enclosure must be
> # supported by the Linux SES driver for this to work.
> #
> ZED_USE_ENCLOSURE_LEDS=1
> ```
> 
> Maybe switch to:
> 
> ```
> # Turn on Fault LED for DEGRADED/FAULTED vdevs
> ZED_USE_ENCLOSURE_FAULT_LEDS=1
> 
> # Turn on Locate LED for OFFLINE/REMOVED/UNAVAIL vdevs
> ZED_USE_ENCLOSURE_LOCATE_LEDS=1
> 
> # Same as ZED_USE_ENCLOSURE_FAULT_LEDS (Deprecated - do not use)
> ZED_USE_ENCLOSURE_LEDS=1
> ```

That would make sense as there seems to be wacky enclosures with only one LED out there. I really like the idea of using the possibility to light both LEDs to indicate a state though.

Any thought on behaviour in the only-one-led case? Light it for all cases?

>     3. Please squash your commits

Will do when we are in agreement on whether this is a good idea at all (ie when this PR is good enough to be moved out of draft state). It's much easier to back up a few steps with the commits there if I misunderstand something and need to redo it some other way...

gdevenyi commented:
> Any thought on behaviour in the only-one-led case? Light it for all cases?
I would say something like a bitmask so you can specify one-or-more of the LEDs lit for any of the cases

tonyhutter commented:
@ZNikke if this PR is of use to you, I think you should go forward and squash your commits and I'll take a look at the code in detail.  Along with that, I think you should make the fault/locate LEDs individually configurable in `zed.rc` (since I think our site will most likely not want to use the locate LEDs):
```bash
# Turn on Fault LED for DEGRADED/FAULTED vdevs
ZED_USE_ENCLOSURE_FAULT_LEDS=1

# Turn on Locate LED for OFFLINE/REMOVED/UNAVAIL vdevs
ZED_USE_ENCLOSURE_LOCATE_LEDS=1

# Same as ZED_USE_ENCLOSURE_FAULT_LEDS (Deprecated - do not use)
ZED_USE_ENCLOSURE_LEDS=1
```

ZNikke commented:
@tonyhutter Updated, squashed and pushed now that I found a few moments to do it. Do note that this is still not tested on hardware, and what I implemented wrt configuration differs from what I think you suggested in behaviour (if one led disabled, uses the remaining led to indicate all states instead of just not reporting those states).

However, I'm really in no hurry to get this one reviewed, as I depend on issue #11950 being fixed first so I can test it properly...

tonyhutter commented:
> However, I'm really in no hurry to get this one reviewed, as I depend on issue #11950 being fixed first so I can test it properly...

FYI - the fix for #11950 was merged yesterday.
